Add generic conversion traits

aturon opened the pull request:
This commit:

* Introduces `std::convert`, providing an implementation of
RFC 529.

* Deprecates the `AsPath`, `AsOsStr`, and `IntoBytes` traits, all
in favor of the corresponding generic conversion traits.

  Consequently, various IO APIs now take `As<Path>` rather than
`AsPath`, and so on. Since the types provided by `std` implement both
traits, this should cause relatively little breakage.

* Deprecates many `from_foo` constructors in favor of `from`.

* Changes `PathBuf::new` to take no argument (creating an empty buffer,
  as per convention). The previous behavior is now available as
  `PathBuf::from`.

* De-stabilizes `IntoCow`. It's not clear whether we need this separate trait.

Closes #14433
Closes #22751

[breaking-change]

r? @alexcrichton 
cc @Gankro @erickt @cmr

aturon commented:
**PRELIMINARY PR**: Do not merge until the RFC is approved.

aturon commented:
cc @sfackler @wycats

aturon commented:
Per IRC: @alexcrichton is very uncomfortable with these living in `libcollections` rather than `libcore`. This placement was to facilitate impls like `impl<T: Clone> To<Vec<T>> for [T]` which would otherwise be disallowed due to coherence.

aturon commented:
cc @reem

aturon commented:
Per IRC: @reem points out that the same coherence problems make it much more difficult to define "implicit conversions" via generics targetting user-land types: you cannot implement `To<MyType<T>> for StdType<T>` due to coherence.

Introducing an ad hoc conversion trait, on the other hand, makes it possible to do this in userland.

aturon commented:
Updated: I've now moved the traits into `libcore`, and have a (rather sad) impl of `To<Vec<u8>> for [u8]` which coherence permits.

bors commented:
:umbrella: The latest upstream changes (presumably #23548) made this pull request unmergeable. Please resolve the merge conflicts.

aturon commented:
I believe this is now ready to go, once the RFC itself has been merged.

alexcrichton commented on src/libcollections/borrow.rs:
```diff
@@ -14,6 +14,7 @@
 
 use core::clone::Clone;
 use core::cmp::{Eq, Ord, Ordering, PartialEq, PartialOrd};
+use core::convert::As;
```
Could these traits also go into the libcore prelude? I think files like this can have `use core::prelude::*` at the top as well (slowly been migrating over time).

alexcrichton commented on src/libcore/convert.rs:
```diff
@@ -0,0 +1,146 @@
+/// A cheap, reference-to-reference conversion.
+pub trait As<T: ?Sized> {
+    /// Perform the conversion.
+    fn convert_as(&self) -> &T;
```
I do quite like your suggestions on the RFC of `as_ref`, `as_mut`, `to`, and `into`. The only one that's "wrong" is `as_ref` (e.g. it should be called `as`) but it seems worth it!

Gankro replied:
Wouldn't as_ref and as_mut conflict with several adhoc methods (thinking of RawPtr and Option off the top of my head)?

aturon replied:
They would overlap, but inherent impls trump trait ones. (Still, the overlap is one reason we might not want to choose these names, or might consider renaming the others...)

alexcrichton commented on src/libcore/convert.rs:
```diff
@@ -0,0 +1,146 @@
+    fn convert_as(&self) -> &[T] {
+        self.as_slice()
+    }
+}
```
I thought we've been pretty hesitant about this in the past? Slices from `Option` and `Result` I thought would continue to be unstable functionality.

aturon replied:
I'm willing to mark this unstable, but personally I have no objection to this functionality, *especially* in the context of generic conversions.

alexcrichton replied:
My fear in marking these unstable is that there's no warning on usage of an unstable *impl*, only on an unstable trait, and it's not clear that we'll get it fixed for 1.0.

alexcrichton commented on src/libcore/option.rs:
```diff
@@ -701,6 +702,19 @@ impl<T> Option<T> {
+
+    /// Convert from `Option<T>` to `&[T]` (without copying)
+    #[inline]
+    #[stable(feature = "rust1", since = "1.0.0")]
```
I would personally prefer this continue to remain unstable

alexcrichton replied:
(same for `Result`)

alexcrichton replied:
ping, perhaps this could remain unstable?

alexcrichton commented on src/librustc/session/search_paths.rs:
```diff
@@ -8,6 +8,7 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
+use std::convert::From;
```
This should be in the prelude, right?

alexcrichton replied:
(same for a few files below)

aturon replied:
It's in the prelude, yes; I probably added a few redundant imports mechanically.

alexcrichton replied:
(ping)

alexcrichton commented on src/libstd/ffi/os_str.rs:
```diff
@@ -98,8 +101,36 @@ impl OsString {
+}
+
+#[stable(feature = "rust1", since = "1.0.0")]
+impl From<String> for OsString {
```
I wonder if these would be better expressed as `Into`? For example functions may want to take `T: Into<OsString>`.

alexcrichton commented on src/libstd/path.rs:
```diff
@@ -999,16 +1036,58 @@ impl PathBuf {
 
 #[stable(feature = "rust1", since = "1.0.0")]
-impl<P: AsPath> iter::FromIterator<P> for PathBuf {
+impl<'a> From<&'a str> for PathBuf {
```
Similarly to the `OsString` case, these may be better expressed via `Into`.

It's also a little sad that the same set of impls as `From<T> for OsString` needs to be listed here, but not necessarily the end of the world.

aturon replied:
Yes, the inability to do blanket "transportation" of one conversion to another is a loss here.

Note that specialization may someday allow us to regain this via automatic reflexive and transitive blankets. It's not clear.

alexcrichton replied:
Yeah I'm willing to just understand that this is a fact of life :)

Gankro commented on src/compiletest/header.rs:
```diff
@@ -324,10 +324,10 @@ fn parse_exec_env(line: &str) -> Option<(String, String)> {
       None => {
         if parse_name_directive(line, "pp-exact") {
-            testfile.file_name().map(|s| PathBuf::new(s))
+            testfile.file_name().map(|s| PathBuf::from(s))
```
This can just be map(PathBuf::from)

Gankro commented on src/libcollections/slice.rs:
```diff
@@ -1088,23 +1089,23 @@ pub trait SliceConcatExt<T: ?Sized, U> {
         for v in self {
             if first { first = false } else { result.push(sep.clone()) }
-            result.push_all(v.as_slice())
+            result.push_all(v.convert_as())
```
This is kind of an unfortunate readability loss.

aturon replied:
Yeah, it's probably more clear to perform the conversion up-front and rebind the variable (perhaps with type annotation, even).

alexcrichton replied:
Note, though, that `v.as_ref::<[_]>()` would probably still be readable.

Gankro replied:
Alternatively if these traits were hooked into coercions you could just `result.push_all(&v)`, which is less informative but somehow less cryptic than `covert_as::<???>()`.

alexcrichton commented:
cc https://github.com/rust-lang/rust/issues/23567

aturon commented:
Updated:

* Removed `To` for the time being.

* Renamed `As` to `AsRef`.

* Renamed all methods to match their trait names.

* All conversion functionality now in `libcore`; added to `core::prelude`

* Added `From` => `Into` implementation, which makes it possible to add conversions in both directions without running afoul of coherence. For example, we now have `From<[T]> for Vec<T>` where `T: Clone`, which yields the corresponding `Into` going in the other direction -- despite the fact that the two types live in different crates.

  I also believe this addresses a few concerns about things implementing `From` instead of `Into`

* Removed `AsRef` impls for slices from `Option`/`Result`, which was causing a bad interaction with their `as_ref` methods. This functionality is available by calling `as_slice` directly.

All in all, I feel *much* better with the form this PR is taking now. It feels cleaner, leaner, more consistent and more extensible. Thanks for the feedback making this possible!

@alexcrichton: re-r?

alexcrichton commented on src/libstd/env.rs:
```diff
@@ -67,8 +67,8 @@ pub fn current_dir() -> io::Result<PathBuf> {
 #[stable(feature = "env", since = "1.0.0")]
-pub fn set_current_dir<P: AsPath + ?Sized>(p: &P) -> io::Result<()> {
-    os_imp::chdir(p.as_path())
+pub fn set_current_dir<P: AsRef<path::Path> + ?Sized>(p: &P) -> io::Result<()> {
```
Could this module import `Path` directly and use `AsRef<Path>`?

aturon replied:
Ahha! It can now. I think when I first wrote it `Path` was still in the prelude :)

alexcrichton commented:
I, too, am quite happy with how this is turning out! r=me with the `as_slice` methods on `Option` and `Result` remaining `#[unstable]` (as they were before)

aturon commented:
pings addressed :)

alexcrichton commented:
@bors: r+ 8389253

aturon commented:
@bors: r=alexcrichton 97c8e43

alexcrichton commented:
Rolled into #23654

bors commented:
:umbrella: The latest upstream changes (presumably #23654) made this pull request unmergeable. Please resolve the merge conflicts.
